ts/src/config.ts

```diff
@@ -171,6 +171,11 @@ export interface ProfilerConfig extends GoogleAuthOptions {
   disableSourceMaps: boolean;
 }
 
+// Interface for an initialized profiler config.
+export interface ProfilerConfig extends LocalConfig {
+  projectId: string;
```

Comment:
Apologies if this was discussed earlier, and I missed it. Why are we requiring all users, even those not running on Istio to now start providing a `projectId`, when their code works fine as-is? With this breaking change, their code may stop compiling. In some cases they would need to go though significant hoops to provide a project id, wherein they didn't need to before.

Looking at the implementation, we actually are able to fetch the `projectId` from the metadata service on the users behalf, and use it even if the `projectId` was not specified in the config. IOW, this breaking change seems more strict than what the code actually requires.

Am I overlooking something?

Reply:
We always required project ID. It should be either discoverable from metadata server, provided in a config file or provided as a init config. Previously, the code attempted to read project ID from metadata server on every profile upload. This causes a problem with Istio env. This PR should make the agent init wait till it can obtain the project ID from metadata server or fail. This should not break any existing code but will require a small change if users are already running on Istio.

Reply:
As additional reassurance that this is working as intended, our Kokoro system test uses a typescript benchmark that doesn't specify the project ID when starting profiler, and it continues to build and pass.

Reply:
Thanks for clarifying that `ProfiledConfig` is not publicly exported to the users, but rather the user-visible config is `Config`.